packages/connect/src/connector/extension.ts

```diff
@@ -57,95 +50,204 @@ export const createScClient = (): ScClient => {
+    // add-chain message.
+    if (isWellKnown) {
+      postToExtension({
+        origin: "substrate-connect-client",
+        chainId: chainState.id,
+        type: "add-well-known-chain",
+        chainName: chainSpecOrWellKnownName,
+      })
+    } else {
+      postToExtension({
+        origin: "substrate-connect-client",
+        chainId: chainState.id,
         type: "add-chain",
-        chainSpec,
+        chainSpec: chainSpecOrWellKnownName,
+        potentialRelayChainIds,
+      })
+    }
+
+    // Wait for the extension to send back either a confirmation or an error.
+    // Note that `initFinished` becomes ready when `chainState` has been modified. The outcome
+    // can be known by looking into `chainState`.
+    await initFinished
+
+    // In the situation where we tried to create a well-known chain, the extension isn't supposed
+    // to ever return an error. There is however one situation where errors can happen: if the
+    // extension doesn't recognize the desired well-known chain because it uses a different list
+    // of well-known chains than this code. To handle this, we download the chain spec of the
+    // desired well-known chain and try again but this time as a non-well-known chain.
+    if (isWellKnown && chainState.state.state === "dead") {
+      // Note that we keep the same id for the chain for convenience.
+      let resolve: undefined | (() => void)
+      const initFinished = new Promise((res) => {
+        resolve = () => res(null)
+      })
+      chainState.state = {
+        state: "wait-ack-or-err",
+        waitFinished: resolve!,
+      }
+
+      postToExtension({
+        origin: "substrate-connect-client",
+        chainId: chainState.id,
+        type: "add-chain",
+        chainSpec: await getSpec(chainSpecOrWellKnownName),
         potentialRelayChainIds: [],
       })
+
+      await initFinished
     }
 
+    // Now check the `chainState` to know if things have succeeded.
+    if (chainState.state.state === "dead") {
+      throw chainState.state.error
+    }
+
+    // Everything is successful.
     const chain: Chain = {
       sendJsonRpc: (jsonRpcMessage) => {
-        if (crashError) throw crashError
-        if (!chains.has(chain)) throw new AlreadyDestroyedError()
+        if (chainState.state.state === "dead") {
+          throw chainState.state.error
+        }
+
         if (!jsonRpcCallback) throw new JsonRpcDisabledError()
-        postToExtension({ type: "rpc", jsonRpcMessage })
+        postToExtension({
+          origin: "substrate-connect-client",
+          chainId: chainState.id,
+          type: "rpc",
+          jsonRpcMessage,
+        })
       },
       remove: () => {
-        if (crashError) throw crashError
-        if (!chains.has(chain)) throw new AlreadyDestroyedError()
-        listeners.delete(chainId)
+        if (chainState.state.state === "dead") {
+          throw chainState.state.error
+        }
```

Comment:
This code is multiple times, maybe we should add it in a function that will accept the state as an input?

Reply:
I mean, it's 3 lines of code that are present 3 times.

Reply:
Its not important. Just a proposal